### DISCUSSION TITLE
Sigstore support in PowerShell

### DISCUSSION THREAD

**Discussion by landon-lengyel** [PowerShell/PowerShell#23988] [Enhancements]
Hi, 

Would it be feasible to add native [Sigstore ](https://www.sigstore.dev/) signing support to PowerShell? This seems like a reasonable alternative to expensive code signing certs for people who aren't making money with their development (like me). However it would require some native PowerShell support to look up the validity of the certificates to be useful.

Could be a nice improvement to the PowerShell Gallary ecosystem though. Glancing through I can find many popular unsigned certificates.

Thank you.

**Comment by rhubarb-geek-nz** [comment]
Hi, interesting idea. Reading [Sigstore](https://www.sigstore.dev/) at appears that the idea is that you can get a code signing certificate that is valid for ten minutes.

It sounds like the code-signing equivalent of Let's Encrypt where the only thing it does is enable encryption, not assert any credence to the actual site. Likewise `Sigstore` would allow the signature to be verified but makes no claims over the validity of the publisher.

Existing code-signing certificates don't just cost money they also require a basic vetting regarding the identity of the publisher.

As it is, Get-AuthenticodeSignature will already validate a signature


```
SignerCertificate                         Status                              StatusMessage                       Path
-----------------                         ------                              -------------                       ----
601A8B683F791E51F647D34AD102C38DA4DDB65F  Valid                               Signature verified.                 test2.ps1

```
or

```
SignerCertificate                         Status                              StatusMessage                       Path
-----------------                         ------                              -------------                       ----
601A8B683F791E51F647D34AD102C38DA4DDB65F  HashMismatch                        The contents of file C:\Users\…     test2.ps1
```

So you can happily sign your scripts and validate if you have installed the CAs. I have my own self-signed CA that I issue my own code-signing certificates with.

Its a bit rough trying to run pwsh enforcing signatures itself when it does not trust Microsoft.

```
C:\Users\rhubarb-geek-nz\tmp>pwsh -ExecutionPolicy AllSigned
PowerShell 7.2.21
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.


Do you want to run software from this untrusted publisher?
File C:\program files\powershell\7\Modules\PSReadLine\PSReadLine.format.ps1xml is published by CN=Microsoft Corporation, O=Microsoft Corporation,
L=Redmond, S=Washington, C=US and is not trusted on your system. Only run scripts from trusted publishers.
[V] Never run  [D] Do not run  [R] Run once  [A] Always run  [?] Help (default is "D"): 
```

As a cross platform user authenticode is not supported on anything other than windows so is not particularly important.

The trust model for installing modules is to trust the repository rather than the package. The recommended way of publishing is to use [Publish-Module](https://learn.microsoft.com/en-us/powershell/module/powershellget/publish-module?view=powershellget-3.x), I don't see a way of signing a NuGet package with that method.

**Comment by landon-lengyel** [comment]
That's an interesting point that Nuget would have to support it as well. I had not considered that.

I decided against using a self-signed cert/CA for the module I'm preparing to publish since I don't believe it would remove the warnings of running an untrusted publisher's code. However, I did not know that you can't even fully run pwsh with AllSigned enabled - I got a chuckle out of that. If that is the case, perhaps self-signed certs are the preferred way to go? It would at least prove that the patches are coming from me, just not prove my true identity.

If that is the preferred way for PowerShell, there may still be value in that Sigstore would likely be easier to use for people than spinning up their own CA and trying to maintain it.
